### PR TITLE
Parallelize precompiles generation II

### PR DESCRIPTION
**Pull request by petvana**
This follows up #47983 and introduce asynchronous read for repl precompiles. In contrast to https://github.com/IanButterworth/julia/commit/5ee671ad3c40533a3b3965f8bcbce20d18014cbf?w=1 from @IanButterworth , it uses `readavailable` and `BufferStream` to handle end of lines correctly. It saves only a few seconds, but improves the fancy printing for impatient developers.

### CODE REVIEW

**Comment by IanButterworth** [comment]
> It saves only a few seconds

For me the gains are bigger

before #47983 
```
Generating REPL precompile statements... 40/40
Executing precompile statements... 1756/1760
Precompilation complete. Summary:
Generation ──  77.218106 seconds 79.4016%
Execution ───  20.031916 seconds 20.5984%
Total ───────  97.250022 seconds
```
#47983 
```
Collecting and executing precompile statements
└ Collect (Basic: ✓ 730, REPL 40/40: ✓ 1659) => Execute ✓ 1751
Precompilation complete. Summary:
Total ───────  80.336407 seconds
```

This PR
```
Collecting and executing precompile statements
└ Collect (Basic: ✓ 730, REPL 40/40: ✓ 1657) => Execute ✓ 1750
Precompilation complete. Summary:
Total ───────  64.902187 seconds
```

File: contrib/generate_precompile.jl [hunk 1]
```diff
@@ -388,48 +389,66 @@ generate_precompile_statements() = try # Make sure `ansi_enablecursor` is printe
...
+        buffer_reader = @async for statement in eachline(precompile_copy)
+            print_state("step2" => "R$n_step2")
             push!(statements_step2, statement)
             n_step2 += 1
         end
+
+        open(precompile_file, "r") do io
+            while true
+                # We need to allways call eof(io) for bytesavailable(io) to work
+                eof(io) && istaskdone(repl_inputter) && eof(io) && break
```
Comment (IanButterworth) [inline contrib/generate_precompile.jl:442]: Just checking, are you saying both eof checks are needed here?

Comment (petvana) [reply contrib/generate_precompile.jl:442]: Yes, it seems so. The first updates the status of `io` not to stuck at `readavailable(io)`. The second is for correctness. We should first wait for the task to finish and then for the EOF.

**Comment by petvana** [comment]
> For me the gains are bigger

Half a minute, as advertised :wink:

**Comment by IanButterworth** [comment]
Seems good to me. 

Does the switch to serial mode for debugging with stdout rather than devnull still work?

**Comment by petvana** [comment]
> Does the switch to serial mode for debugging with stdout rather than devnull still work?

Not sure, what you mean by `stdout rather than devnull`, but the switch still works and those 3 steps are done in serial order.

**Comment by IanButterworth** [comment]
I just grouped all the debug options together at the top of the file to make it clearer. Seems to work for me, if you disable pretty printing, disable parallel, and enable repl input viewing

**Review by IanButterworth — APPROVED** [review]
Seems good to me, but might be good to have a @vtjnash review too

**Comment by IanButterworth** [comment]
Out of curiosity I reordered `repl_script` to have Profile then Pkg then the ones in this file and saw 5 seconds faster. 

Not sure if it's worth doing as it breaks the ordering (could be a followup PR), but at least it makes sense that doing the larger workloads first helps parallelization
```
Collecting and executing precompile statements
└ Collect (Basic: ✓ 730, REPL 40/40: ✓ 1661) => Execute ✓ 1753
Precompilation complete. Summary:
Total ───────  60.710451 seconds
```

**Comment by giordano** [comment]
> what you mean by `stdout rather than devnull`

https://github.com/JuliaLang/julia/blob/cc5bc87b005ee193bd22b2907d591f120371a26b/contrib/generate_precompile.jl#L289

**Comment by petvana** [comment]
> Out of curiosity I reordered `repl_script` to have Profile then Pkg then the ones in this file and saw 5 seconds faster.

I'm kind of surprised, it saved you so much time. I believe, we can reorder repl script as we want because it will remain deterministic.

**Comment by IanButterworth** [comment]
ok! Pushed

**Comment by IanButterworth** [comment]
Windows failures are unrelated. @vtjnash does this look ok to you?

**Comment by IanButterworth** [comment]
I'll merge as this appears to work as designed. If @vtjnash has any issue with the buffering design that can be addressed in a followup

**Comment by maleadt** [comment]
I just had a build hang (wasn't using the changes from this PR yet):

```
Collecting and executing precompile statements
└ Collect (Basic: ✓ 730, REPL 39/40: ◒ ) => Execute ◒ 494^C^C^C
```

As there was no CPU usage, but the main process was still happily spinning, I suspect this may be related to these parallel precompilation changes. The process also wasn't interruptible, i.e., CTRL-C is ignored (as you can see in the output). Are either of these issues known?

**Comment by StefanKarpinski** [comment]
We really need to get better at making ctrl-C work. I know that @vtjnash will tell me that it can't be made to work completely reliably, which, sucks, sure, but it can be made to work much better than it does and we should do that.

**Comment by IanButterworth** [comment]
@maleadt if you do hit it again, can you SIGINFO/SIGUSR1 profile it to see what it's doing. That's probably more reliable than getting a trace from an interrupt

**Comment by petvana** [comment]
> I just had a build hang (wasn't using the changes from this PR yet):

I've noticed a hang when building without stdlibs, i.e., `stdlibs = []` in `sysimg.jl`. It is reproduceble, but happened even before #47983. Thankfully, this can be fixed easily by disabling the whole REPL collection as `have_repl = false`.

**Comment by vtjnash** [comment]
LGTM. Copying streams (as done here) is one of the best actual use cases for `readavailable`

**Pull request merged**